File: src/my_project/interviews/top_150_questions_round_22/ex_89_course_schedule_ii.py
from typing import List, Union, Collection, Mapping, Optional
from abc import ABC, abstractmethod
from collections import deque, defaultdict

class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        # Build adjacency list and in-degree array
        graph = defaultdict(list)
        in_degree = [0] * numCourses
        
        for course, prereq in prerequisites:
            graph[prereq].append(course)
            in_degree[course] += 1
        
        # Initialize queue with courses that have no prerequisites
        queue = deque([i for i in range(numCourses) if in_degree[i] == 0])
        result = []
        
        # Process courses in topological order
        while queue:
            course = queue.popleft()
            result.append(course)
            
            # Reduce in-degree for dependent courses
            for next_course in graph[course]:
                in_degree[next_course] -= 1
                if in_degree[next_course] == 0:
                    queue.append(next_course)
        
        # If we processed all courses, return the order; otherwise return empty list
        return result if len(result) == numCourses else []

File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_89_course_schedule_ii.ts
function findOrder(numCourses: number, prerequisites: number[][]): number[] {
    // Build adjacency list and in-degree array
    const graph: Map<number, number[]> = new Map();
    const inDegree: number[] = new Array(numCourses).fill(0);
    
    // Initialize graph
    for (let i = 0; i < numCourses; i++) {
        graph.set(i, []);
    }
    
    // Build the graph
    for (const [course, prereq] of prerequisites) {
        graph.get(prereq)!.push(course);
        inDegree[course]++;
    }
    
    // Initialize queue with courses that have no prerequisites
    const queue: number[] = [];
    for (let i = 0; i < numCourses; i++) {
        if (inDegree[i] === 0) {
            queue.push(i);
        }
    }
    
    const result: number[] = [];
    
    // Process courses in topological order
    while (queue.length > 0) {
        const course = queue.shift()!;
        result.push(course);
        
        // Reduce in-degree for dependent courses
        for (const nextCourse of graph.get(course)!) {
            inDegree[nextCourse]--;
            if (inDegree[nextCourse] === 0) {
                queue.push(nextCourse);
            }
        }
    }
    
    // If we processed all courses, return the order; otherwise return empty array
    return result.length === numCourses ? result : [];
}

  